src/transport/MessageCounter.h

```diff
@@ -0,0 +1,123 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2020 Google LLC.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP message counters.
+ *
+ */
+#pragma once
+
+#include <support/PersistedCounter.h>
+
+namespace chip {
+
+/**
+ * MessageCounter represents a local message counter. There are 3 types
+ * of message counter
+ *
+ * 1. Global unencrypted message counter
+ * 2. Global encrypted message counter
+ * 3. Session message counter
+ *
+ * There will be separate implementations for each type
+ */
+class MessageCounter
+{
+public:
+    enum Type : uint8_t
+    {
+        GlobalUnencrypted,
+        GlobalEncrypted,
+        Session,
+    };
+
+    virtual ~MessageCounter() = 0;
+
+    virtual Type GetType()     = 0;
+    virtual void Reset()       = 0;
+    virtual uint32_t Value()   = 0; /** Get current value */
+    virtual uint32_t Advance() = 0; /** Advance the counter and get new value */
+};
+
+inline MessageCounter::~MessageCounter() {}
+
+class GlobalUnencryptedMessageCounter : public MessageCounter
+{
+public:
+    GlobalUnencryptedMessageCounter();
+    ~GlobalUnencryptedMessageCounter() override {}
+
+    Type GetType() override { return GlobalUnencrypted; }
+    void Reset() override
+    { /* null op */
+    }
+    uint32_t Value() override { return value; }
+    uint32_t Advance() override { return ++value; }
+
+private:
+    uint32_t value;
+};
+
+class GlobalEncryptedMessageCounter : public MessageCounter
+{
+public:
+    GlobalEncryptedMessageCounter() {}
+    ~GlobalEncryptedMessageCounter() override {}
+
+    CHIP_ERROR Init();
+    Type GetType() override { return GlobalEncrypted; }
+    void Reset() override
+    { /* null op */
+    }
+    uint32_t Value() override { return persisted.GetValue(); }
+    uint32_t Advance() override { return static_cast<uint32_t>(persisted.Advance()); }
```

Comment:
This ... does not look right.  It's going to cast the CHIP_ERROR return value of `PersistedCounter::Advance` to an integer, which will presumably nearly-always return 0.  Is this tested at all? Or are we only testing the fake shim, as opposed to the code that will actually run?

Reply:
Oops, fixed.